File: Graph/547. Number of Provinces/numberOfProvinces.js
/**
 * @param {number[][]} isConnected
 * @return {number}
 */
 var findCircleNum = function(isConnected) {
    let n = isConnected.length;
    let output = 0;
    let visited = new Set();
    
    for (let city = 0; city < n; city++) {
        if (!visited.has(city)) {
            output += 1;
            dfs(city);
        }
    }
    
    return output;
    
    function dfs(city) {
        visited.add(city);
        
        for (let sister = 0; sister < n; sister++) {
            if (isConnected[city][sister]
                && !visited.has(sister))
                dfs(sister);
        }
    }
};

File: Graph/547. Number of Provinces/numberOfProvinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        def dfs(city):
            visited.add(city)
            for sister in range(n):
                # if the cities are connected
                # and not in visited
                # check on the sister city
                if isConnected[city][sister] and sister not in visited:
                    dfs(sister)
        
        n = len(isConnected)
        output = 0
        visited = set()
       
        for city in range(n):
            # if the city not in visisted
            # it is a start for a new province
            # bring this city to dfs and check its sisters
            if city not in visited:
                output += 1
                dfs(city)
        
        return output